tests/infra_test.go

```diff
@@ -121,19 +126,77 @@ var _ = Describe("Infrastructure", func() {
 			pod := pods.Items[0] // only one compute node in the test environment
 
 			Eventually(func() string {
-				stdout, _, err := tests.ExecuteCommandOnPodV2(virtClient,
-					&pod, "virt-handler",
-					[]string{
-						"curl",
-						"-L",
-						"-k",
-						fmt.Sprintf("https://%s:%s/metrics", pod.Status.PodIP, "8443"),
-					})
-				Expect(err).ToNot(HaveOccurred())
-				lines := filterMetricsOutput(stdout, "kubevirt")
+				out := getKubevirtVMMetrics(virtClient, &pod, "virt-handler")
+				lines := takeMetricsWithPrefix(out, "kubevirt")
 				return strings.Join(lines, "\n")
 			}, 30*time.Second, 2*time.Second).Should(ContainSubstring("kubevirt"))
 		}, 300)
+
+		It("should include the storage metrics for a running VM", func() {
+			By("Creating the VirtualMachineInstance")
+
+			// WARNING: we assume the VM will have a VirtIO disk (vda)
+			// and we add our own vdb on which we do our test.
+			// but if the default disk is not vda, the test will break
+			// TODO: introspect the VMI and get the device name of this
+			// block device?
+			vmi := tests.NewRandomVMIWithEphemeralDisk(tests.ContainerDiskFor(tests.ContainerDiskAlpine))
+			tests.AppendEmptyDisk(vmi, "testdisk", "virtio", "1Gi")
+
+			By("Starting a new VirtualMachineInstance")
+			obj, err := virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(tests.NamespaceTestDefault).Body(vmi).Do().Get()
+			Expect(err).ToNot(HaveOccurred(), "Should create VMI")
+
+			By("Waiting until the VM is ready")
+			tests.WaitForSuccessfulVMIStart(obj)
+
+			By("Expecting the VirtualMachineInstance console")
+			expecter, err := tests.LoggedInAlpineExpecter(vmi)
+
+			Expect(err).ToNot(HaveOccurred())
+			defer expecter.Close()
+
+			By("Writing some data to the disk")
+			_, err = expecter.ExpectBatch([]expect.Batcher{
+				&expect.BSnd{S: "dd if=/dev/zero of=/dev/vdb bs=1M count=1\n"},
+				&expect.BExp{R: "localhost:~#"},
+				&expect.BSnd{S: "sync\n"},
+				&expect.BExp{R: "localhost:~#"},
+			}, 10*time.Second)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Finding the prometheus endpoint")
+			l, err := labels.Parse("kubevirt.io=virt-handler")
+			Expect(err).ToNot(HaveOccurred())
+			pods, err := virtClient.CoreV1().Pods(tests.KubeVirtInstallNamespace).List(metav1.ListOptions{LabelSelector: l.String()})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(pods.Items).ToNot(BeEmpty())
+
+			By("Scraping the Prometheus endpoint")
+			pod := pods.Items[0] // only one compute node in the test environment
+
+			var lines []string
+			for tryNum := 0; tryNum < 20; tryNum++ {
```

Comment:
Why is that loop needed? Is there a reason to believe that the metrics will not be scraped?

Reply:
This loop is needed to handle retries. There are timing issues in the tests, I just don't want to give up after the first failed attempt. 20 tries is overshoot by high margin, I expect success within the first 2-3 tries in the common case, 5-6 worst (from what I've seen so far)

Reply:
I see. Maybe using `Eventually` would be more appropriate than this extra loop?